### PR TITLE
Dropdown new guidelines

### PR DESCRIPTION
**Pull request by nicl**
In general I am really pleased with the result of applying our new stricter approach. There are a few questions/issues which I've flagged in comments and would especially be good to get thoughts on.

## What does this change?

Converts dropdown to new style guidelines.

## Why?

For consistency and simplicity.

### CODE REVIEW

File: packages/guui/dropdown/index.js [hunk 1]
```diff
@@ -20,131 +20,139 @@ type Props = {
...
-            top: '14px',
-            left: '10px',
-            width: '10px',
-            height: '4px',
-            transform: 'rotate(-45deg)',
-        },
-    },
-});
+const input = css`
+    // TODO re-add screen-reader only mixin
```
Comment (nicl) [inline packages/guui/dropdown/index.js]: At the moment the screenReaderOnly export is defined as a simple object. We should probably rewrite this as a css export but I wasn't sure if to do that here or separate PR.

File: packages/guui/dropdown/index.js [hunk 2]
```diff
@@ -228,20 +241,22 @@ export default class Dropdown extends Component<
...
-                        <li key={link.title}>
+                    {links.map((l, index) => (
+                        <li key={l.title}>
                             <a
-                                href={link.url}
-                                className={link.isActive ? 'active' : ''}
+                                href={l.url}
+                                className={`${l.isActive ? linkActive : link} ${
+                                    index === 0 ? linkFirst : ''
+                                }`}
```
Comment (nicl) [inline packages/guui/dropdown/index.js]: The bit above is a bit fiddly and currently breaks our 'one class per element' rule. Any thoughts on how to improve this welcome!

File: packages/guui/index.js [hunk 3]
```diff
@@ -7,7 +7,7 @@ import { extractCritical } from 'emotion-server';
 import React, { Component } from 'react';
 import { hydrate as hydrateApp, render } from 'react-dom';
 import { renderToString as reactRenderToString } from 'react-dom/server';
-import styled from 'react-emotion';
+import styled, { css } from 'react-emotion';
```
Comment (nicl) [inline packages/guui/index.js]: I know the plan is to use these directly but I've hacked this here (and in my other PRs) for expediency just so I could test stuff.

File: packages/guui/dropdown/index.js (hunk 1 above)
Comment (GHaberis) [reply packages/guui/dropdown/index.js]: Ah yeah interesting, I guess the export and all the places it's used would need to be modified in a single pull request, unless we export an Object syntax version of `screenReaderOnly` and a TTL version temporarily and remove the object once all dependants are up to date.

File: packages/guui/dropdown/index.js [hunk 4]
```diff
@@ -20,131 +21,139 @@ type Props = {
...
+    transition: color 80ms ease-out;
+    padding: 6px 10px;
+    margin: -1px 0 0 0;
+    text-decoration: none;
+
+    display: block;
+    padding: 10px 18px 15px 30px;
+
+    :hover {
+        text-decoration: underline;
```
Comment (GHaberis) [inline packages/guui/dropdown/index.js]: No need for this rule `text-decoration: underline` as it's being overridden 2 lines later with `text-decoration: none;`

File: packages/guui/dropdown/index.js (hunk 2 above)
Comment (GHaberis) [reply packages/guui/dropdown/index.js]: I'm wondering whether the single className rule is a premature optimisation that shifts the inefficiencies elsewhere (duplicating style rules across multiple classes). I also think it could lead to fiddly issues like this where we have to write more JS logic to pick the correct single classname.

File: packages/guui/dropdown/index.js (hunk 4 above)
Comment (nicl) [reply packages/guui/dropdown/index.js]: ah thanks!

File: packages/guui/dropdown/index.js (hunk 2 above)
Comment (nicl) [reply packages/guui/dropdown/index.js]: Hey, yeah I'm not sure. It's not too hard to rewrite it as a function:

    const linkStyle = (isActive, isFirst) => {
        if (isActive && isFirst) {
            return activeAndFIrst
        } else if (isFirst) {
            return firstLink
        } else (isActive) {
            return activeLink
        }

        return normalLink
    }

Performance-wise emotion seems to be pretty clear that generating styles dynamically is bad - so the idea is to pre-declare them like above.

Comment (GHaberis) [reply packages/guui/dropdown/index.js]: Hey, my preference right now is multiple class names, eg:

```JS
<Link className={`baseLinkStyles ${isActive ? 'activeLinkStyles': ''} ${isFirst ? 'firstLinkStyles': ''}`}>Click me</Link>
```
There are libs like [classNames](https://www.npmjs.com/package/classnames) that can be used for conditionally creating classNames too.

I'd say the benefits of this are:
- Less duplication of style rules within the source code, if you want to change the  "base" rule of link you update it once.
- Less duplication of style rules within generated CSS
- More css-like to combine styles this way

Comment (SiAdcock) [reply packages/guui/dropdown/index.js]: After some discussion we've concluded that [`cx`](https://github.com/emotion-js/emotion/blob/master/docs/cx.md), the Emotion-specific implementation of `classNames` generates the most performant CSS in this case.

```js
<Link className={
  cx({
    [link]: true,
    [linkActive]: isActive, 
    [linkFirst] isFirst,
  })
>Click me</Link>
```

File: packages/guui/components/Dropdown/index.js [hunk 5]
```diff
@@ -4,10 +4,10 @@
  */
 
 import { Component } from 'react';
-import styled from 'react-emotion';
+import { css } from 'react-emotion';
 import palette from '@guardian/pasteup/palette';
 import { textSans } from '@guardian/pasteup/fonts';
-import { screenReaderOnly } from '@guardian/pasteup/mixins';
+// import { screenReaderOnly } from '@guardian/pasteup/mixins';
```
Comment (GHaberis) [inline packages/guui/components/Dropdown/index.js]: Can we remove this commented out import?

File: packages/guui/components/Dropdown/index.js [hunk 6]
```diff
@@ -21,131 +20,138 @@ type Props = {
...
+        transform: translateY(-2px) rotate(45deg);
+        border: 1px solid currentColor;
+        border-left: transparent;
+        border-top: transparent;
+        margin-left: 6px;
+        vertical-align: middle;
+        transition: transform 250ms ease-out;
+    }
+
+    .expanded:after,
```
Comment (GHaberis) [inline packages/guui/components/Dropdown/index.js]: This `.expanded:after` block can be removed

File: frontend/components/Header/Nav/EditionDropdown.js [hunk 7]
```diff
@@ -32,8 +32,8 @@ export default () => {
     ];
 
     return (
-        <EditionDropdown>
+        <div className={editionDropdown}>
```
Comment (AWare) [inline frontend/components/Header/Nav/EditionDropdown.js]: Our previous use of `styled()` was causing problems for the React dev tools, but now that we're not- we should probably not `export default () => <JSX goes here/>` as this shows up as Unknown. `const Name = ()=> <JSX/>; export default const;` will show us Name in the tool.

File: packages/guui/components/Dropdown/index.js [hunk 8]
```diff
@@ -193,18 +199,21 @@ export default class Dropdown extends Component<
...
 
         return (
-            <Div>
+            <div>
                 {this.state.noJS ? (
                     [
                         <label
                             htmlFor={checkboxID}
-                            className={this.state.isExpanded ? 'expanded' : ''}
+                            className={
```
Comment (GHaberis) [inline packages/guui/components/Dropdown/index.js]: We've discussed elsewhere such as the[ incoming style rules](https://github.com/guardian/dotcom-rendering/commit/6b57464) that we should be using `cx` to conditionally render classNames. This `button` like a good use case for  using cx to conditionally apply class names here. For example...

```js
const button = css`
    color: blue;
    font-weight: 700;
`;

const expanded = css`
    color: red;
`;

const Button = ({ isExpanded }) =>
    <button className={cx({
        button: true,
        expanded: isExpanded
    )}}>Click Me</a>;
```

File: packages/guui/components/Dropdown/index.js [hunk 9]
```diff
@@ -219,7 +228,9 @@ export default class Dropdown extends Component<
                 ) : (
                     <button
                         onClick={this.toggle_}
-                        className={this.state.isExpanded ? 'expanded' : ''}
+                        className={
```
Comment (GHaberis) [inline packages/guui/components/Dropdown/index.js]: As per my previous comment, this looks like a use case for using `cx` to conditionally render class names.

File: packages/guui/components/Dropdown/index.js [hunk 10]
```diff
@@ -229,20 +240,22 @@ export default class Dropdown extends Component<
...
                 >
-                    {links.map(link => (
-                        <li key={link.title}>
+                    {links.map((l, index) => (
+                        <li key={l.title}>
                             <a
-                                href={link.url}
-                                className={link.isActive ? 'active' : ''}
+                                href={l.url}
+                                className={`${l.isActive ? linkActive : link} ${
```
Comment (GHaberis) [inline packages/guui/components/Dropdown/index.js]: As per my previous comment, this looks like a use case for using `cx` to conditionally render class names.

File: packages/guui/components/Dropdown/index.js (hunk 8 above)
Comment (AWare) [reply packages/guui/components/Dropdown/index.js]: I just tried this in my PR and AFAICT it'll need to be

```
const Button = ({ isExpanded }) =>
    <button className={cx({
        [button]: true,
        [expanded]: isExpanded
    )}}>Click Me</a>;
```

 or

 ```
const Button = ({ isExpanded }) =>
    <button className={cx(
        button,{
        [expanded]: isExpanded
    )}}>Click Me</a>;
```
I prefer the second form. I think the benefit is more in terms of standardising how and where logic in applying styles should be applied and belong. I think at this level of complexity, ternarys would be fine- but they'd start to lose out with any greater complexity.

Comment (nicl) [reply packages/guui/components/Dropdown/index.js]: Hey, thanks and completely agree that cx pays off as things get more complex. It just feels like what you've written is, for the current use case at least - which is quite a common one, a lot less readable than what's currently there.

File: packages/guui/components/Dropdown/index.js (hunk 10 above)
Comment (nicl) [reply packages/guui/components/Dropdown/index.js]: Definitely agree that double-ternary as here is not great.

File: packages/guui/components/Dropdown/index.js (hunk 8 above)
Comment (SiAdcock) [reply packages/guui/components/Dropdown/index.js]: As I understand it, there's a performance reason for using `cx` instead of the ternary. 

With the ternary, styles for both `button` and `buttonExpanded` would always be shipped, regardless of whether `isExpanded` is true or false. `buttonExpanded` is essentially all the styles in `button` duplicated, plus a few extras.

With `cx`, only styles for `button` and `buttonExpanded` (which in our case would contain no duplication, just the styles `:after` pseudo-selector) would be shipped if is `isExpanded` was false.

If this is difficult to visualise, you can try playing around with this [codesandbox](https://codesandbox.io/s/w2z5mzoxol). But as a rule of thumb, I think it's worth assuming that `cx` _always tries to optimise your CSS output_, whereas ternaries do not.

Comment (nicl) [reply packages/guui/components/Dropdown/index.js]: Ah that's helpful. It makes sense. I guess that in this case there is the change of having to generate styles at runtime/client-side but that feels okay.
